File: RemixedDungeon/src/main/java/com/nyrds/pixeldungeon/mobs/guts/SuspiciousRat.java
package com.nyrds.pixeldungeon.mobs.guts;

import com.nyrds.Packable;
import com.nyrds.pixeldungeon.ml.R;
import com.nyrds.platform.audio.Sample;
import com.nyrds.platform.util.StringsManager;
import com.watabou.pixeldungeon.Assets;
import com.watabou.pixeldungeon.actors.Char;
import com.watabou.pixeldungeon.actors.CharUtils;
import com.watabou.pixeldungeon.actors.blobs.ToxicGas;
import com.watabou.pixeldungeon.actors.mobs.Mob;
import com.watabou.pixeldungeon.sprites.CharSprite;
import com.watabou.pixeldungeon.utils.GLog;

import org.jetbrains.annotations.NotNull;

/**
 * Created by DeadDie on 12.02.2016
 */
public class SuspiciousRat extends Mob {

	private static final float TIME_TO_HATCH = 4f;

	{
		hp(ht(140));
		baseDefenseSkill = 25;
		baseAttackSkill  = 25;
		dmgMin = 10;
		dmgMax = 15;
		dr = 2;

		expForKill = 1;
		maxLvl = 30;

		pacified = true;

		addImmunity(ToxicGas.class);
	}

	@Packable
	private boolean transforming = false;

	@Override
	public boolean act() {
		if (enemySeen) {
			if (!transforming) {
				spend(TIME_TO_HATCH);
				transforming = true;
				if (CharUtils.isVisible(this)) {
                    getSprite().showStatus(CharSprite.NEGATIVE, StringsManager.getVar(R.string.Goo_StaInfo1));
                    GLog.n(StringsManager.getVar(R.string.SuspiciousRat_Info1));
				}
				getSprite().zap(getEnemy().getPos());
			} else {
				int wereratPos = this.getPos();
				if (level().cellValid(wereratPos)) {
					PseudoRat mob = new PseudoRat();
					mob.setPos(wereratPos);
					level().spawnMob(mob);
					Sample.INSTANCE.play(Assets.SND_CURSED);
				}
				die(this);
			}
			return true;
		}
		return super.act();
	}

	@Override
	public boolean zap(@NotNull Char enemy){
		return false;
	}
}

File: scripts/mobs/SuspiciousRat.lua
local RPD = require "scripts/lib/commonClasses"
local mob = require "scripts/lib/mob"
-- java SuspiciousRat (batch 12): once it sees the enemy it twitches, then
-- becomes a wererat (PseudoRat) ~4 ticks later. Java spent 4 ticks at once;
-- the act hook cannot spend, so the delay is a tick counter - same wall clock.
local TRANSFORM_TICKS = 4

return mob.init{
    act = function(self)
        if not self.enemySeen then
            return
        end
        local data = mob.restoreData(self)
        if not data.transforming then
            data.transforming = true
            data.ticks = 0
            self:showStatus(RPD.CharSprite.NEGATIVE, RPD.textById("Goo_StaInfo1"))
            RPD.glogn(RPD.textById("SuspiciousRat_Info1"))
            local enemy = self:getEnemy()
            if enemy ~= nil then
                self:getSprite():zap(enemy:getPos())
            end
            return
        end
        data.ticks = data.ticks + 1
        if data.ticks < TRANSFORM_TICKS then
            return
        end
        local pos = self:getPos()
        if RPD.Dungeon.level:cellValid(pos) then
            local wererat = RPD.MobFactory:mobByName(RPD.MobFactory.PSEUDO_RAT)
            wererat:setPos(pos)
            RPD.Dungeon.level:spawnMob(wererat)
            RPD.playSound("snd_cursed")
        end
        self:die(self)
    end
}
